<img src="img/Marca-ITBA-Color-ALTA.png" width="200">


# Programación para el Análisis de Datos

## Clase 2 - funciones, clases y objetos

### Objetivo

En esta clase se analizarán los conceptos básicos para entender el funcionamiento y las ventajas de la utilización de funciones, clases y objetos en la programación.

### Funciones

Una función es un bloque de código reutilizable que puede ser ejecutado por el interprete mediante el nombre de la función.

### Definición
Para definir una función en python es necesario utilizar la palabra reservada (keyword) ***def*** y respetar la sintaxis.

``` python
def nombre_funcion():
    "Codigo de la función"
    ...
```
Una función puede ejecutar cualquier bloque de código contenido. El código de la función esta compuesto por todas las líneas tabuladas posteriores a ***def***.

In [1]:
def mi_funcion():
    # Principio de la función
    print("Esta función esta siendo ejecutada")
    a = 1
    b = a + 10
    print(a , b)
    # Fin de la funcion

In [2]:
# Para ejecutar la función es necesario invocarla
mi_funcion()

Esta función esta siendo ejecutada
1 11


### Argumentos

Las funciones permiten utilizar variables externas mediante argumentos. 
Existen distintas formas de definir argumentos de una función.
La más simple es mediante **argumentos obligatorios**

``` python
def mostrar_valores(a, b):
    print("a = {}\nb= {}".format(a, b))
```

Este tipo de argumentos son obligatorios para que la función pueda ejecutarse.
Es importante tener en cuenta que, en este caso, la función considera el orden de los argumentos.



In [3]:
def mostrar_valores(a, b):
    print("primera variable = {}\nsegunda variable = {}".format(a, b))

mostrar_valores(1, [3, 2])

primera variable = 1
segunda variable = [3, 2]


In [4]:
c = 'hola'
d = 9

mostrar_valores(c, d)

primera variable = hola
segunda variable = 9


Si la cantidad de argumentos dados es erronea se produce una excepción:

In [5]:
try:
    mostrar_valores(1)
except Exception as e:
    print(e)

mostrar_valores() missing 1 required positional argument: 'b'


Python permite utilizar el nombre de los argumentos para invocar una función. Si bien esto no es necesario, es recomendable aplicar esta práctica, ya que facilita la lectura del código.

En el siguiente ejemplo utilizaremos la función que definimos anteriormente.

In [6]:
print("Caso 1")
mostrar_valores(a=1, b=[3, 2])

Caso 1
primera variable = 1
segunda variable = [3, 2]


In [7]:
print("Caso 2")
mostrar_valores(b=1, a=[3, 2])

Caso 2
primera variable = [3, 2]
segunda variable = 1


In [8]:
print("Caso 3")
mostrar_valores(1, b=[3, 2])

Caso 3
primera variable = 1
segunda variable = [3, 2]


In [9]:
# Python considera que el primer argumento corresponde al primer argumento posicional
print("Caso 4")
try:
    mostrar_valores(1, a=[3, 2])
except Exception as e:
    print(e)
    print("El interprete de Python no realiza suposiciones a la hora de asignar argumentos")

Caso 4
mostrar_valores() got multiple values for argument 'a'
El interprete de Python no realiza suposiciones a la hora de asignar argumentos


Es necesario tener en cuenta que a la hora de invocar una función, python no permite utilizar argumentos sin nombre luego de haber utilizado un argumento con nombre (**keyword argument**)

``` pyhton
mostrar_valores(b=1, [3, 2])
```
SyntaxError: non-keyword arg after keyword arg

Esto se considera un error de sintaxis.

#### Valores por defecto

Python permite asignar valores por defecto a los argumentos. De esta forma se pueden definir argumentos opcionales.


In [10]:
def mi_funcion(a=1, b=10, mostrar_suma=False):
    print("a = {}\nb = {}".format(a, b))
    if mostrar_suma:
        print("a + b = {}".format(a + b))

In [11]:
print("Caso 1")
mi_funcion()

Caso 1
a = 1
b = 10


In [12]:
print("Caso 2")
mi_funcion(11)

Caso 2
a = 11
b = 10


In [13]:
print("Caso 3")
mi_funcion(b=15)

Caso 3
a = 1
b = 15


In [14]:
print("Caso 4")
mi_funcion(3, 4, True)

Caso 4
a = 3
b = 4
a + b = 7


#### Argumentos de largo variable

En ocasiones, una función tiene una cantidad de argumentos desconocida o no siempre recibe los mismos argumentos.
En estos casos se pueden utilizar argumentos de largo variable ( **Variable-length arguments**)

Estos argumentos se definen de la siguiente forma 
``` python 
def my_func([argumentos_fijos,] *argumentos_variables ):
    "Muestra por pantalla los argumentos fijos"
    print("Los argumentos fijos son: {}".format(argumentos_fijos)
    "Muestra por pantalla, si existen, los argumentos variables"
    for var in argumentos_variables:
          print(var)
    return None
```
Cuando se antepone un asterisco al nombre del argumento, el interprete entiende que este es un argumento de tamaño variable.

A medida que se agregan argumentos cuando se llama la función, estos se van almacenando en la variable **argumentos_variables** de forma secuencial.

In [15]:
def my_func(arg1 , *argvar):
    print("El argumento fijo vale: {}".format(arg1))
    print("El argumento variable vale: {}".format(argvar))
    print("type argumento variable: {}".format(type(argvar)))
    for var in argvar:
          print(var)
    return None
          


In [16]:
my_func(1, 2, 3, 4, 5)

El argumento fijo vale: 1
El argumento variable vale: (2, 3, 4, 5)
type argumento variable: <class 'tuple'>
2
3
4
5


In [17]:
def my_func(arg1 , **kargvar):
    print("El argumento fijo vale: {}".format(arg1))
    print("El argumento variable vale: {}".format(kargvar))
    print("type argumento variable: {}".format(type(kargvar)))
    return None
    

In [18]:
      
my_func(1, segundo=2, tercero=3, cuarto=4)

El argumento fijo vale: 1
El argumento variable vale: {'segundo': 2, 'tercero': 3, 'cuarto': 4}
type argumento variable: <class 'dict'>


In [19]:
my_func(segundo=2, tercero=3, cuarto=4, arg1=1)

El argumento fijo vale: 1
El argumento variable vale: {'segundo': 2, 'tercero': 3, 'cuarto': 4}
type argumento variable: <class 'dict'>


### Return

Las funciones de python tienen la capacidad de procesar datos y devolver un resultado (o varios).

En la clase anterior ya hemos utilizado funciones que retornan un valor, como por ejemplo la funcion *len()*.

Para darle este comportamiento a nuestra función se utiliza la palabra clave **return**.

In [20]:
from random import randint

def func_return():
    var1 = "String1"
    var2 = randint(0,1000)
    if var2 > 500:
        return var1
    else: 
        return var2
    print("Este codigo no se ejecuta")
    


In [21]:
for i in range(5):
    print(func_return())


String1
240
36
String1
432


En el anterior ejemplo se puede ver que se puede utilizar la keyword **return** en multiples ocasiones, pero ésta se ejecutará únicamente una vez por llamada.
Cuando el interprete lee *return*, retorna ese resultado al programa que ejecutó la función.

</br>
Al igual que la asignación multiple, es posible retornar múltiples variables


In [22]:
def func():
    a = 10
    return a, a+1, a**2

print("La funcion retorna una variable de tipo: {}".format(type(func())))

La funcion retorna una variable de tipo: <class 'tuple'>


In [23]:
# La tupla se puede asignar a múltiples variables:
v1, v2, v3 = func()
print("v1 = {}\nv2 = {}\nv3 = {}".format(v1, v2, v3))


v1 = 10
v2 = 11
v3 = 100


In [24]:
v1, *v2 = func()

print(v1, v2)
print(v1, *v2)

10 [11, 100]
10 11 100


### Funciones anónimas

Una función anónima es simplemente una función que no tiene nombre.
Estas funciones son mayormente utilizadas cuando una función es aplicada únicamente en un contexto particular. 
El caso de uso más común es utilizarlo con la función *map* y la función *filter*.

#### Sintaxis

``` python
lambda arguments: expression
```

Las funciones anónimas son comunmente conocidas como funciones lambda. Esto se debe a que en multiples lenguajes de programación se utiliza esta palabra clave.

Este tipo de funciones pueden contener únicamente una línea de código y retornan un valor.

Una función anónima se puede asignar a una variable y utilizarse como una función.

<!-- Averiguar si se puede sobreescribir para agregarle a las listas el map o el filter -->

In [25]:
doble = lambda x : x * 2
n = 5
print("doble es de tipo {}\nEl resultado de doble({}) es {}".format(type(doble), n, doble(n)))

doble es de tipo <class 'function'>
El resultado de doble(5) es 10


la función doble tambien se podría definir como:
``` python
def doble(x):
    return x * 2
```
Ambas definiciones son muy similares.
Algunas diferencias entre una función y una función lambda son:
- Las funciones lambda pueden contener únicamente una linea de código

#### Ejemplos

In [26]:
# Un programa que obtiene el triple de sus elementos

lista = [1, 5, 4, 6, 8, 11, 3, 12]

nueva_lista = list(map(lambda x: x * 3 , lista))

print(nueva_lista)

[3, 15, 12, 18, 24, 33, 9, 36]


In [27]:
def triple(x):
    return 3*x

list(map(triple, lista))

[3, 15, 12, 18, 24, 33, 9, 36]

In [28]:
# Un programa que obtiene los multiplos de 8

lista = range(100)

nueva_lista = list(filter(lambda x: x%8 == 0 , lista))

print(nueva_lista)

[0, 8, 16, 24, 32, 40, 48, 56, 64, 72, 80, 88, 96]


### NameSpace

Python utiliza distintos espacios de memoria para almacenar variables y funciones

Un NameSpace es un espacio de memoria donde se almacenan nombres de variables y su correspondiente ubicación en la memoria.

En python hay 4 namespaces:
- Built-In
- Global
- Enclosing
- Local

![namespace](img/namespace.png)

#### Built-in
Este namespace contiene todas las variables integradas a python por defecto. Este namespace se crea cuando se ejecuta el interprete y está presente en todo momento de la ejecución.

Para ver las funciones presentes en este namespace es necesario ejecutar el siguiente código:
``` python
dir(__builtins__)
```

#### Global
El namespace global corresponde a la variables definidas por el usuario dentro de la rama principal del programa (main)

```python
def a():
    pass

b = 1
```

En el ejemplo, *a* y *b* pertenecen al namespace *global*

Python a su vez define un namespace global para cada uno de los módulos que importa. Para acceder a ese namespace se utiliza la siguiente sintaxis:
nombreModulo.variableOFuncion

Por ejemplo:

In [29]:
import math

sin = lambda x: x

x1 = sin(math.pi)
x2 = math.sin(math.pi)

print("""En el namespace 'math' la funcion sin es el seno de un numero real.
        En el namespace 'global' sin representa una funcion identidad (el valor que retorna es el valor de entrada)
        x1 = {}
        x2 = {}""".format(x1, x2))

En el namespace 'math' la funcion sin es el seno de un numero real.
        En el namespace 'global' sin representa una funcion identidad (el valor que retorna es el valor de entrada)
        x1 = 3.141592653589793
        x2 = 1.2246467991473532e-16


Para ver los nombres definidos en el Namespace *global*, se puede utilizar la función `globals()`

``` python
print(globals())

globals()
```

#### Enclosing y Local
Cada vez que el interprete de python ejecuta una función, se crea un namespace.

Es posible definir funciones adentro de funciones. 

Por ejemplo
``` python
def f():
  print('Empieza f()')
  
  def g():
    print('Empieza g()')
    print('Termina g()')
    return
  g()
  print('Termina f()')
  return

f()
```

El resultado de este bloque sería:
``` bash
Empieza f()
Empieza g()
Termina g()
Termina f()
```

Empieza la ejecución del bloque (namespace *'main'*), se llama a la función f() (Se crea un namespace *'Enclosing'*). Luego f() llama a la función g() (se genera un namespace (*'local'* adentro del namespace de f())

Cada uno de los namespaces existe hasta que finaliza la ejecución del mismo. 
En el caso de una función, cuando finaliza el bloque o se ejecuta un *return*.

Para mostrar los nombres definidos en el namespace local se utiliza la funcion `locals()` que retorna un diccionario con los nombres y valores de las variables definidas.

In [30]:
def f(x, y):
    s = 'foo'
    l = [1, 2]
    print(locals())

f(10, 0.5)

{'x': 10, 'y': 0.5, 's': 'foo', 'l': [1, 2]}


#### Scope de una variable

En Python es importante entender en que namespace existe una variable. Esto es importante ya que se podría utilizar el mismo nombre para una variable y almacenar distintos valores.

Por ejemplo, continuando con el módulo *math*:

In [31]:
import math
# math.pi = 3.14159...
pi = 3.14

def getPi():
    pi = 3.141
    print(pi)
    
    
print(math.pi)
print(pi)
getPi()
print(globals()['pi'])

3.141592653589793
3.14
3.141
3.14


En el ejemplo, la variable pi tiene 3 valores distintos. En el modulo math es 3.14159265359, en el entorno principal ( o main) *pi* vale 3.14 y dentro de la función *getPi* vale 3.141

Python utiliza la regla LEGB. Esto significa que el interpreta va a buscar la variable primero en el namespace *Local*, luego en el namespace *Enclosing*, luego en *Global* y por ultimo en *Built-In*. En caso de no encontrar el nombre buscado, el interprete eleva una Excepción del tipo *NameError exception*



### Módulos

Un módulo es un objeto de python que agrupa código. Este código puede contener variables y funciones.

Dentro de la carpeta *src* se encuentra un archivo **module.py** que contiene el código en python mostrado a continuación.

``` python

print("Este es el modulo 'module.py'")

a = 10

b = 20


def f1(x, y):
    print("esta función se encuentra en el modulo 'module'")
    return x, y

```

Para importar el archivo se utiliza la palabra clave **import**

Al importarse un módulo, es ejecutado el código contenido en el archivo *.py* y se genera un namespace propio para ese módulo.

En el siguiente bloque de código se importa el módulo analizado.

In [32]:
import src.module as m

print(m.a)

m.f1(1,2)

Este es el modulo 'module.py'
10
Esta función se encuentra en el modulo 'module'


(1, 2)

Es importante conocer la ubicación del módulo para que el interprete sea capaz de encontrarlo. 
Los módulos instalados mediante conda y pip, se pueden importar directamente, ya que su ubicación se encuentra en el **path** del interprete.

Al módulo importado se le puede dar un alias con la keyWord **as**, esto es útil cuando el nombre del módulo es largo.

Los módulos se pueden estructurar en paquetes, como en este caso **src**. Los paquetes son una forma de estructurar el espacio de nombres de los módulos de Python utilizando la sintaxis de puntos (paquete.subpaquete.modulo).

Además, tenemos a las librerías. Una librería es una colección de módulos y/o paquetes que proporcionan funcionalidades para un propósito específico o resuelven un conjunto particular de problemas. A menudo, las librerías son desarrolladas y distribuidas por terceros. Una librería consta de uno o más módulos y/o paquetes. Ejemplos de librerías son numpy, pandas, tensorflow, etc. 


También es posible importar unicamente algunas funciones del módulo al namespace del programa (**global**)


In [33]:
from src.module import f1, suma

f1(1,2)

print("La suma de 1 y 2 es: {}".format(suma(1,2)))

Esta función se encuentra en el modulo 'module'
La suma de 1 y 2 es: 3


## Clases

Una clase es un bloque de código que nos permite agrupar funciones, variables y constantes. Este blóque de código es facilmente accesible y se puede replicar la cantidad de veces que sea necesaria.

Agrupar código nos permite desarrollar un código legible y mantenible.

### Definición

Para definir una clase se utiliza la palabra reservada **Class**

``` python
class Operaciones: 
    # Código de la clase
    def suma(self, a, b):
        return a + b
    
    def producto(self, a, b):
        return a * b
    
    def cuadrado(self, a):
        return pow(a,2)
```

El código que se encuentra en una clase es similar al que vimos en el módulo, pero se pueden ver algunas diferencias.

Se puede ver que en cada función se utiliza como primer argumento la palabra reservada **self**. Esta keyword representa a la instancia de clase y es utilizada para acceder a las funciones y variables de la misma.

Veamos ahora cómo utilizar una clase en el código.
``` python
o = Operaciones()

o.suma(2, 3)
## 5

p = Operaciones()
p.producto(2, 3)
## 6
```

Se puede ver que una clase se asigna a una variable.  Esto se conoce como **instanciar**. Esa variable tendrá todas las funcionalidades de la clase, como por ejemplo las funciones y los atributos.

Entonces se podria decir que la principal diferencia entre un módulo y una clase es que para utilizar el código de una clase hay que **instanciarla**. El resultado de esto es un **objeto**.

En el código se puede ver que una misma clase se puede instanciar varias veces.

### Instancia

Una instancia es, entonces, una variable que contiene todas las funcionalidades de una clase.

Para instanciar una clase se utiliza el nombre de la clase como función. 

Es una buena practica nombrar las clases con una mayuscula al inicio del nombre.

In [34]:
class Operaciones: 
    # Código de la clase mínimo
    pass
    
o = Operaciones()

Cuando se instancia una clase python llama a una funcion predefinida:  **\_\_init\_\_**
Esta función se puede sobreescribir para agregar las funcionalidades necesarias.

En el siguiente ejemplo agregamos la funcion **\_\_init\_\_** al código.

In [35]:
class Operaciones: 
    # Definicion de __init__
    def __init__(self):
        print("La clase Operaciones ha sido instanciada")
    

In [36]:
o = Operaciones()

La clase Operaciones ha sido instanciada


La función de inicialización tambien puede tener argumentos específicos.

In [37]:
class Operaciones: 
    def __init__(self, num):
        print("La clase Operaciones ha sido inicializada con {}".format(num))
    
o = Operaciones(1)

La clase Operaciones ha sido inicializada con 1


In [38]:
p = Operaciones(2)

La clase Operaciones ha sido inicializada con 2


### Variables

Al igual que en los módulos, las clases pueden contener variables propias. Estas variables suelen conocerce con el nombre de atributos.

#### Variables de instancia
Un atributo de instancia es definido en la función \_\_init\_\_. Cada instancia de la clase tendrá este atributo y podrá ser modificado sin alterar el valor en otras instancias.


In [39]:
class Operaciones: 
    def __init__(self, num):
        self.num = num


In [40]:
o = Operaciones(2)
print("El atributo 'num' vale: {}".format(o.num))


El atributo 'num' vale: 2


In [41]:
o.num = 5
print("El atributo 'num' vale: {}".format(o.num))


El atributo 'num' vale: 5


In [42]:
p = Operaciones(3)

p.num

3

#### Variables de clase

Las variables de clase son comunes a todas las instancias de la clase y a la clase.
Es buena práctica definir todas las variables de clase al inicio de la misma.

A continuación se muestra un ejemplo de un atributo de clase:

In [43]:
class Operaciones: 
    class_attr = 1
    def __init__(self, num):
        self.inst_attr = num
        
o = Operaciones(2)

print("El atributo 'class_attr' vale: {}".format(o.class_attr))

print("El atributo 'class_attr' está definido para la clase: {}".format(Operaciones.class_attr))


El atributo 'class_attr' vale: 1
El atributo 'class_attr' está definido para la clase: 1


### Métodos
Para agregar funcionalidades a la clase, se pueden agregar funciones a las clases. Estas funciones son conocidas como métodos.
Una clase puede contener todos los métodos que se desee.
El método debe contener siempre el argumento *self* en la primera posición. 

#### Métodos comunes
Python define algunos métodos que son comunes a todas las clases y son útiles para interactuar con funciones específicas de python. Éstos son conocidos como **magic methods**.
Estos métodos se caracterizan por estar rodeados por **\_\_**.

Algunos ejemplos son:
- \_\_init\_\_ : inicializa la instancia
- \_\_call\_\_ : permite utilizar la clase como una función
- \_\_srt\_\_ : esta función se ejecuta cuando se hace un print() de la clase
- \_\_contains\_\_ : esta función se llama cuando se utiliza un operador de pertenencia como *in*

Para mas información: [link](https://rszalski.github.io/magicmethods/#appendix1)

En data science con frecuencia vamos a trabajar con clases que implementan algún determinado algoritmo, por ejemplo la regresión lineal. Cuando lleguemos a los módulos de Machine Learning y empecemos a trabajar con la librería `Scikit-Learn`, va a ser práctica común instanciar clases de modelos. 

Para tener una mejor comprensión de lo que vamos a estar haciendo al trabajar con clases de modelos, definamos una clase muy sencilla: 

Definimos la clase `Modelo`

En esta clase podemos pasar parámetros al momento de instaciar la clase, que serán inicializados por el método `init`, mostar los párametros por pantalla y ejecutar 

In [44]:
class Modelo:
    def __init__(self, parametro1, parametro2, modelo):
        self.p1 = parametro1
        self.p2 = parametro2
        self.model = modelo
    
    def __str__(self):
        return "El modelo es {}, con parámetros: p1= {} y p2= {}".format(self.model, self.p1, self.p2)
    
    def __call__(self, x):
        if self.model == 'lineal':
            return self.p1*x + self.p2 
        
        elif self.model == 'exponencial':
            return self.p2*(self.p1**x)
        
        else:
            print('Error: modelo no válido')
      

In [45]:
      
modelo1 = Modelo(2, 5, 'lineal')

print(modelo1)

print(modelo1(1))

El modelo es lineal, con parámetros: p1= 2 y p2= 5
7


In [46]:
type(modelo1)

__main__.Modelo

In [47]:
modelo2 = Modelo(2, 5, 'exponencial')

print(modelo2)

modelo2(1)

El modelo es exponencial, con parámetros: p1= 2 y p2= 5


10

#### Métodos personalizados

Una de las mayores ventajas de utilizar métodos personalizados es agregar funcionalidades y crear clases más intuitivas y fáciles de usar.

Volviendo al ejemplo anterior, se puede definir un método *predict* que realice la predición.

Si bien, el comportamiento es el mismo, es mas intuitivo que el método tenga el nombre predict

In [48]:
class Modelo:
    def __init__(self, parametro1, parametro2, modelo):
        self.p1 = parametro1
        self.p2 = parametro2
        self.model = modelo
    
    def __str__(self):
        return "El modelo es {}, con parámetros: p1= {} y p2= {}".format(self.model, self.p1, self.p2)
    
    def predict(self, x):
        if self.model == 'lineal':
            return self.p1*x + self.p2 
        
        elif self.model == 'exponencial':
            return self.p2*(self.p1**x)
        
        else:
            print('Error: modelo no válido')
 

In [49]:
modelo1 = Modelo(2, 5, 'exponencial')

print(modelo1)

modelo1.predict(1)

El modelo es exponencial, con parámetros: p1= 2 y p2= 5


10

In [50]:
modelo1.predict(2)

20

Algunos errores comunes corresponden al uso de *namespaces*. Cada clase que se instancia, genera un namespace local.
Analicemos el siguiente ejemplo:

In [51]:
class Modelo2:
    def __init__(self, parametro1):
        self.p1 = parametro1
    
    def __str__(self):
        return "El modelo tiene parámetro: p1 = {}".format(self.p1)
    
    def predict(self, x):
        return p1 + x
    
p1 = 4


In [52]:

modelo2 = Modelo2(2)
print(modelo2)

El modelo tiene parámetro: p1 = 2


In [53]:
modelo2.predict(1)

5

Se puede ver que la predicción no tiene el valor esperado.
Viendo el código, se puede ver que predict retorna *p1 + x*
``` python
def predict(self, x):
        return p1 + x
```
Al no utilizar la keyword *self*, el interprete busca la variable en un namespace superior. En este caso encuentra el nombre en el namespace *global* con un valor de *4*


### Herencia

La **herencia** es un mecanismo de la programación orientada a objetos que sirve para crear clases nuevas a partir de clases preexistentes. Se toman (heredan) atributos y comportamientos de las clases viejas y se los modifica para modelar una nueva situación.

La clase vieja se llama **clase base** o **superclase** y la que se construye a partir de ella es una **clase derivada** o **subclase**.

#### Herencia implicita

Creamos 2 clases, la base y la derivada.


In [54]:
class BaseImpl:
    
    def implicito(self):
        print("Método implicito() de Base")
        
        
class DerivadaImpl(BaseImpl):
    pass

Instanciamos ambas clases, creando los objetos madre e hija.

In [55]:
madre_impl = BaseImpl()
hija_impl = DerivadaImpl()

Ejecutamos el método implicita() en el objeto madre de la clase base

In [56]:
madre_impl.implicito()

Método implicito() de Base


Ejecutamos el método implicita() en el objeto hija de la clase derivada

In [57]:
hija_impl.implicito()

Método implicito() de Base


A pesar de no haber definido una función `implicita()` en la clase Derivada, puedo llamar el método en el objeto hijo. Esto muestra que todas las funciones que están en una clase base van a estar en las clases derivadas.

#### Sobrescritura explícita

Creamos 2 clases, la base y la derivada.

In [58]:
class BaseSobr:
    
    def sobrescritura(self):
        print("Método sobrescritura() de Base")
        
        
class DerivadaSobr(BaseSobr):
    
    def sobrescritura(self):
        print("Método sobrescritura() de Derivada")

In [59]:
madre_sobr = BaseSobr()
hija_sobr = DerivadaSobr()

In [60]:
madre_sobr.sobrescritura()

Método sobrescritura() de Base


El método sobrescritura() de la clase derivada reemplaza a la de la clase base:

In [61]:
hija_sobr.sobrescritura()

Método sobrescritura() de Derivada


Veamos un ejemplo de clases con herencia. Vamos a definir una clase para crear perfiles de personas.

Creamos a la clase Persona:

In [62]:
class Persona:   
    def __init__(self, nombre, apellido, dni):
        """ Método init para ingresar nombre, apellido y dni al instanciar la clase"""
        self.nombre = nombre
        self.apellido = apellido
        self.dni = dni
        self.lista_estudios = []
    
    def estudios(self, estudio):
        """ Método para agregar estudios a la lista de estudios"""
        self.lista_estudios.append(estudio)
               

Creamos el objeto María a partir de la clase Persona:

In [63]:
maria = Persona('María', 'Lopez', '29.892.382')

maria.estudios('Lic Economia')
maria.estudios('MBA')

In [64]:
print("Nombre: {}".format(maria.nombre))
print("Apellido: {}".format(maria.apellido))
print("DNI: {}".format(maria.dni))
print("Estudios: {}".format(maria.lista_estudios))

Nombre: María
Apellido: Lopez
DNI: 29.892.382
Estudios: ['Lic Economia', 'MBA']


Ahora vamos a definir una subclase que crea alumnos de ITBA. 

Nota: la definición de clase a continuación contiene algunos problemas que abordaremos más adelante. ¿Pueden detectarlos?

In [65]:
class AlumnoITBA(Persona):
    '''Clase que hereda de Persona'''
   
    lista_de_cursos = []
    
    def agregarCurso(self, curso):
        '''Método para agregar cursos'''
        self.lista_de_cursos.append(curso)

In [66]:
pedro = AlumnoITBA('Pedro','Mazzini', '32.564.278')
pedro.estudios('Ing. Industrial')
pedro.agregarCurso('Data Science')

In [67]:
print("Nombre: {}".format(pedro.nombre))
print("Apellido: {}".format(pedro.apellido))
print("DNI: {}".format(pedro.dni))
print("Estudios: {}".format(pedro.lista_estudios))
print("Cursos: {}".format(pedro.lista_de_cursos))

Nombre: Pedro
Apellido: Mazzini
DNI: 32.564.278
Estudios: ['Ing. Industrial']
Cursos: ['Data Science']


Para saber si una clase es subclase de otra, se utiliza la función `issubclass()`.

In [68]:
# Sintaxis: issubclass(<clase_1>, <clase_2>)

issubclass(AlumnoITBA, Persona)

True

In [69]:
issubclass(Persona, AlumnoITBA)

False

¿ Qué pasaría si ITBA necesitara el género, además del nombre, apellido y DNI que nos da la clase Persona?

Por otro lado, al definir la clase AlumnoITBA, nos quedó la lista de cursos como un atributo de clase, lo cual es incorrecto porque los cursos corresponden a cada alumno, es decir a la instancia. Para resolver estos temas usamos la función **super()**.

#### Sobrescritura de métodos con super()

En algunos casos es conveniente reutilizar parte del código de un método de una **superclase** que ha sido sobrescrito por el método de la **subclase**.

La funcion `super()` permite insertar el código del método de la superclase que ha sido sobreescrito.

In [70]:
class AlumnoITBA2(Persona):
    
    def __init__(self, nombre, apellido, dni, genero):
        self.lista_de_cursos = []
        if genero.lower() in ['masculino', 'femenino', 'otro']:
            self.genero = genero
        else:
            raise ValueError
        super().__init__(nombre, apellido, dni)
        
        
    def agregar_curso(self, curso):
        self.lista_de_cursos.append(curso)

In [71]:
laura = AlumnoITBA2('Laura','Lagos','29.369.324','femenino')
laura.estudios('Ing. Industrial')
laura.agregar_curso('Data Engineering')

In [72]:
print("Nombre: {}".format(laura.nombre))
print("Apellido: {}".format(laura.apellido))
print("Género: {}".format(laura.genero))
print("DNI: {}".format(laura.dni))
print("Estudios: {}".format(laura.lista_estudios))
print("Cursos: {}".format(laura.lista_de_cursos))

Nombre: Laura
Apellido: Lagos
Género: femenino
DNI: 29.369.324
Estudios: ['Ing. Industrial']
Cursos: ['Data Engineering']


In [73]:
laura.agregar_curso('Data Science')

In [74]:
laura.lista_de_cursos

['Data Engineering', 'Data Science']

Para ver detalles sobre una clase, así como los métodos que hereda y de cuáles clases, podemos usar la función `help()`.

In [75]:
help(AlumnoITBA2)

Help on class AlumnoITBA2 in module __main__:

class AlumnoITBA2(Persona)
 |  AlumnoITBA2(nombre, apellido, dni, genero)
 |
 |  Method resolution order:
 |      AlumnoITBA2
 |      Persona
 |      builtins.object
 |
 |  Methods defined here:
 |
 |  __init__(self, nombre, apellido, dni, genero)
 |      Método init para ingresar nombre, apellido y dni al instanciar la clase
 |
 |  agregar_curso(self, curso)
 |
 |  ----------------------------------------------------------------------
 |  Methods inherited from Persona:
 |
 |  estudios(self, estudio)
 |      Método para agregar estudios a la lista de estudios
 |
 |  ----------------------------------------------------------------------
 |  Data descriptors inherited from Persona:
 |
 |  __dict__
 |      dictionary for instance variables
 |
 |  __weakref__
 |      list of weak references to the object



Python permite agregar una documentación personalizada utilizando docstrings.

In [76]:
class AlumnoITBA2h(Persona):
    """
        Esta clase permite cargar alumnos del itba. Hereda todas las funcionalidades de la clase Persona
    """
    def __init__(self, nombre, apellido, dni, genero):
        '''Añade el atributo genero al método __init__ de la superclase 
        y definimos la lista de cursos dentro del init'''
        self.lista_de_cursos = []
        if genero.lower() in ['masculino', 'femenino', 'otro']:
            self.genero = genero
        else:
            raise ValueError
        super().__init__(nombre, apellido, dni)
        
        
    def agregar_curso(self, curso):
        '''Método para agregar cursos'''
        self.lista_de_cursos.append(curso)
        
help(AlumnoITBA2h)

Help on class AlumnoITBA2h in module __main__:

class AlumnoITBA2h(Persona)
 |  AlumnoITBA2h(nombre, apellido, dni, genero)
 |
 |  Esta clase permite cargar alumnos del itba. Hereda todas las funcionalidades de la clase Persona
 |
 |  Method resolution order:
 |      AlumnoITBA2h
 |      Persona
 |      builtins.object
 |
 |  Methods defined here:
 |
 |  __init__(self, nombre, apellido, dni, genero)
 |      Añade el atributo genero al método __init__ de la superclase
 |      y definimos la lista de cursos dentro del init
 |
 |  agregar_curso(self, curso)
 |      Método para agregar cursos
 |
 |  ----------------------------------------------------------------------
 |  Methods inherited from Persona:
 |
 |  estudios(self, estudio)
 |      Método para agregar estudios a la lista de estudios
 |
 |  ----------------------------------------------------------------------
 |  Data descriptors inherited from Persona:
 |
 |  __dict__
 |      dictionary for instance variables
 |
 |  __weakref__

En las siguientes clases utilizaremos diversas clases. Es común que los módulos implementen clases para modularizar las funcionalidades.

Por ejemplo el módulo numpy define 'ndarray'. En la siguiente clase veremos en profundidad las funcionalidades de este módulo.


<!-- <span style="font-size:1.5em">Fin de la clase.</span> -->

<span style="font-size:2em">¡Muchas gracias por su atención!</span>